### ISSUE TITLE
get_or_insert_with panics in release builds with nightly but works on stable

### ISSUE THREAD

**Issue by fxdave** [EmbarkStudios/poll-promise#18]
**Describe the bug**
It works well in debug build, but it panics in release build.

**To Reproduce**
https://github.com/fxdave/vonal-rust/blob/5e28d452b268ca5a3fa7f59cb6d78f153d6a944f/src/vonal_daemon/plugins/math/mod.rs

**Logs**
```log
Server started, waiting for clients
1
2
3
4
5
6
7
8
1
2
3
4
5
1
2
3
4
5
1
2
3
4
5
thread 'main' panicked at 'internal error: entered unreachable code', /home/dbiro/.cargo/registry/src/index.crates.io-6f17d22bba15001f/poll-promise-0.2.0/src/promise.rs:332:49
stack backtrace:
   0: rust_begin_unwind
             at /rustc/3a8a131e9509c478ece1c58fe0ea2d49463d2300/library/std/src/panicking.rs:577:5
   1: core::panicking::panic_fmt
             at /rustc/3a8a131e9509c478ece1c58fe0ea2d49463d2300/library/core/src/panicking.rs:67:14
   2: core::panicking::panic
             at /rustc/3a8a131e9509c478ece1c58fe0ea2d49463d2300/library/core/src/panicking.rs:117:5
   3: <vonal::plugins::math::Math as vonal::plugins::Plugin>::search
   4: core::ops::function::FnOnce::call_once{{vtable.shim}}
   5: egui::ui::Ui::allocate_ui_with_layout_dyn
   6: vonal::app::App::render_search_screen
   7: core::ops::function::FnOnce::call_once{{vtable.shim}}
   8: egui::containers::frame::Frame::show
   9: egui::containers::panel::CentralPanel::show
  10: egui::context::Context::run
  11: egui_glow::winit::EguiGlow::run
  12: vonal::handle_platform_event
  13: winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration
  14: winit::platform_impl::platform::x11::EventLoop<T>::run_return
  15: vonal::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Device:**
 - OS: Arch Linux
 - Rust Stable: rustc 1.68.2 (9eb3afe9e 2023-03-27)
 - Rust Nightly: rustc 1.70.0-nightly (28a29282f 2023-04-06)

If you think it's unrelated to the library, please upstream the bug to rustc.

**Comment by zmtq05** [comment]
The same error happened in the stable version(1.70).

CODE:
```rust
fn main() {
    Promise::spawn_thread("a", || ()).block_and_take();
}
```

- OS: Windows 10
- toolchain: stable-x86_64-pc-windows-msvc rustc 1.70.0 (90c541806 2023-05-31)

**Comment by joboet** [comment]
[The code in question](https://docs.rs/poll-promise/0.2.0/src/poll_promise/promise.rs.html#322-329) is unsound, mutating through a shared reference is **never** allowed, not even after casting it to a mutable pointer. This was fixed in #14, but that has not been released yet.

**Comment by avsaase** [comment]
@emilk sorry to ping you but could you create a new release with the fix from #14?

**Comment by emilk** [comment]
@avsaase sorry for the slowness on this one - `poll_promise` 0.3.0 is now released
